compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECanonicalizeNpuDmaCpyNd.cpp

```diff
@@ -0,0 +1,187 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <memory>
+
+#include "iree-amd-aie/IR/AMDAIEDialect.h"
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+
+#define DEBUG_TYPE "iree-amdaie-canonicalize-npu-dma-cpy-nd"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+class AMDAIECanonicalizeNpuDmaCpyNdPass
+    : public impl::AMDAIECanonicalizeNpuDmaCpyNdBase<
+          AMDAIECanonicalizeNpuDmaCpyNdPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AMDAIEDialect>();
+  }
+
+  AMDAIECanonicalizeNpuDmaCpyNdPass() = default;
+  AMDAIECanonicalizeNpuDmaCpyNdPass(
+      const AMDAIECanonicalizeNpuDmaCpyNdPass &){};
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    ModuleOp moduleOp = getOperation();
+    IRRewriter rewriter(context);
+    Attribute zero = rewriter.getIndexAttr(0);
+    Attribute one = rewriter.getIndexAttr(1);
+
+    WalkResult walkResult = moduleOp->walk([&](NpuDmaCpyNdOp dmaOp) {
+      SmallVector<OpFoldResult> srcOffsets = dmaOp.getSourceMixedOffsets();
+      SmallVector<OpFoldResult> srcSizes = dmaOp.getSourceMixedSizes();
+      SmallVector<OpFoldResult> srcStrides = dmaOp.getSourceMixedStrides();
+
+      SmallVector<OpFoldResult> tgtOffsets = dmaOp.getTargetMixedOffsets();
+      SmallVector<OpFoldResult> tgtSizes = dmaOp.getTargetMixedSizes();
+      SmallVector<OpFoldResult> tgtStrides = dmaOp.getTargetMixedStrides();
+
+      bool allRankLessThan4 = srcOffsets.size() <= 4 && srcSizes.size() <= 4 &&
+                              srcStrides.size() <= 4 &&
+                              tgtOffsets.size() <= 4 && tgtSizes.size() <= 4 &&
+                              tgtStrides.size() <= 4;
+      if (!allRankLessThan4) {
+        dmaOp.emitOpError() << " has offsets/sizes/strides with more than 4 "
+                               "elements, unexpected in this pass.";
+        return WalkResult::interrupt();
+      }
+
+      if (dmaOp.getSourceMemorySpaceAsUInt() == 0) {
+        if (!dmaOp.hasSourceAddressing()) {
+          dmaOp.emitOpError()
+              << "has source in L3, but does not have source addressing. "
+                 "Source addressing is required to canonicalize to rank-4 "
+                 "here.";
+          return WalkResult::interrupt();
+        }
+        // Make rank-4.
+        srcOffsets = getPrepended(srcOffsets, zero);
+        srcSizes = getPrepended(srcSizes, one);
+        srcStrides = getPrepended(srcStrides, zero);
+
+        // Ensure that the zero stride is in the outer-most dimension.
```

Comment:
Clarify that if the zero stride is preceded by non-unit dimensions, the walk will be interrupted.